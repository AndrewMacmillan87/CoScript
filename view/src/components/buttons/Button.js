import React from "react";
import PropTypes from "prop-types";

import "./buttons.css";

const Button = ({ onClick, id, classes, value }) => {
    return (
        <button onClick={onClick} id={id} className={classes} type='button'>
            <span className='btn-text'>
                <b>{value}</b>
            </span>
        </button>
    );
};

Button.propTypes = {
    onClick: PropTypes.func,
    classes: PropTypes.string,
    id: PropTypes.string,
    value: PropTypes.string
};

export default Button;
