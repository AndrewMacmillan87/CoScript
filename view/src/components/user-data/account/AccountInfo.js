import React from "react";
import Alert from "../../alert/Alert";
import EditAccountData from "./EditAccountData";

import PropTypes from "prop-types";

const AccountInfo = ({
    user,
    displayEmailWindow,
    emailField,
    cancelEdit,
    onChange,
    onSubmit,
    onClick
}) => {
    const editIcon = require("../../../img/edit.svg");
    return (
        <section className='profile-form-container'>
            {displayEmailWindow && (
                <EditAccountData
                    heading='Update email'
                    field={emailField}
                    name='email'
                    cancelEdit={cancelEdit}
                    onChange={onChange}
                    onSubmit={onSubmit}
                    value='Save'
                />
            )}
            <div className='profile-box'>
                <label className='profile-label' htmlFor='user-email'>
                    Email Address:
                </label>
                <div className='input-container'>
                    <div className='account-box'>
                        <p className='users-email'>{user && user.email}</p>
                    </div>
                    <button className='edit-icon-btn' onClick={e => onClick(e)}>
                        <img
                            id='edit-email-btn'
                            className='edit-input-icon'
                            src={editIcon}
                            alt='edit input icon'
                        />
                    </button>
                </div>
                <Alert />
            </div>
        </section>
    );
};

AccountInfo.propTypes = {
    displayEmailWindow: PropTypes.bool.isRequired,
    onSubmit: PropTypes.func.isRequired,
    onClick: PropTypes.func.isRequired,
    cancelEdit: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    emailField: PropTypes.object
};

export default AccountInfo;
