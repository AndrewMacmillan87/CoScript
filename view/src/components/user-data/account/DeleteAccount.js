import React, { Fragment } from "react";
import Button from "../../buttons/Button";
import EditAccountData from "./EditAccountData";

import PropTypes from "prop-types";

const DeleteAccount = ({
    onClick,
    onChange,
    cancelEdit,
    onDelete,
    deleteField,
    displayDeleteWindow,
    deleteCode,
    deleteCodeIncorrect,
}) => {
    return (
        <Fragment>
            {displayDeleteWindow && (
                <EditAccountData
                    heading='Delete your account?'
                    field={deleteField}
                    name='confirmDeleteCode'
                    cancelEdit={cancelEdit}
                    onChange={onChange}
                    onSubmit={onDelete}
                    value='Delete'
                    deleteAccount='true'
                    deleteCode={deleteCode}
                    deleteCodeIncorrect={deleteCodeIncorrect}
                />
            )}
            <h3 className='delete-heading'>Delete your account?</h3>
            <p>
                <b>Warning</b>
            </p>
            <p className='delete-disclaimer'>
                You cannot undo this action and all documents associated with
                your account will be permanently deleted. Please only delete
                your account after ensuring any work you have saved on the
                system is safely backed up. Be aware that all master copies of
                projects you created will be sent to all of the collaborators on
                those projects when you delete your account
            </p>
            <Button
                id='delete-account-btn'
                classes='small-btn'
                onClick={(e) => onClick(e)}
                value='Delete'
            />
        </Fragment>
    );
};

DeleteAccount.propTypes = {
    user: PropTypes.object,
    onClick: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    cancelEdit: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    displayDeleteWindow: PropTypes.bool.isRequired,
    deleteField: PropTypes.object,
    deleteCode: PropTypes.number,
    deleteCodeIncorrect: PropTypes.bool.isRequired,
};

export default DeleteAccount;
