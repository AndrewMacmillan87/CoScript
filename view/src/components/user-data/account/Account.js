import React from "react";
import { withRouter } from "react-router-dom";
import MobileNav from "../MobileNav";
import SidePanel from "../SidePanel";
import AccountInfo from "./AccountInfo";
import Password from "./Password";
import DeleteAccount from "./DeleteAccount";

// Redux
import { connect } from "react-redux";
import { displayAlert } from "../../../redux/actions/alertpopup";
import { sendPWEmail, verifyPWCode } from "../../../redux/auxiliary/sendEmail";
import { loadUser, deleteUser } from "../../../redux/actions/auth";
import { updateUserEmail } from "../../../redux/actions/account";

import PropTypes from "prop-types";

import "./account.css";

export class Account extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            email: "",
            password: "",
            confirmPassword: "",
            confirmCode: "",
            confirmDeleteCode: "",
            deleteCode: NaN,
            deleteCodeIncorrect: false,
            verifyFailedMsg: "",
            visible: false,
            authenticated: false,
            failed: false,
            displayEmailWindow: false,
            displayDeleteWindow: false,
            loading: false,
        };
        this.displaySpinner = this.displaySpinner.bind(this);
        this.removeSpinner = this.removeSpinner.bind(this);
        this.cancelEdit = this.cancelEdit.bind(this);
        this.hidePassword = this.hidePassword.bind(this);
        this.onClick = this.onClick.bind(this);
        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.sendChangeCode = this.sendChangeCode.bind(this);
        this.submitCode = this.submitCode.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.emailField = React.createRef();
        this.deleteField = React.createRef();
        this.verifyFailed = React.createRef();
    }

    /**
     * Sets the browser tab title and scrolls to the top
     * of the page when the component mounts
     */
    componentDidMount() {
        document.title = "CoScript - Account";
        window.scrollTo(0, 0);
    }

    /**
     * Checks the 'failed' state property to see if a
     * verification code failed to match and sets the
     * 'verifiedFailed' ref textContext to the value of
     * 'verifyFailedMsg' if it did
     */
    componentDidUpdate() {
        const { failed, verifyFailedMsg } = this.state;
        if (failed) {
            this.verifyFailed.current.textContent = verifyFailedMsg;
        }
    }

    /**
     * Sets the 'loading' state property to true to display
     * a loading spinner to the user while loading
     */
    displaySpinner() {
        this.setState({
            loading: true,
        });
    }

    /**
     * Sets the 'loading' state property to false to remove
     * a loading spinner from the UI
     */
    removeSpinner() {
        this.setState({
            loading: false,
        });
    }

    /**
     * Falsifies all state properties that are associated
     * with displaying UI elements to the user so they can
     * edit their account details
     */
    cancelEdit() {
        this.setState({
            displayEmailWindow: false,
            displayDeleteWindow: false,
            deleteCode: NaN,
            deleteCodeIncorrect: false,
        });
    }

    /**
     * Sets the 'visible' state property to false to hide the
     * password update verification code field. This also clears
     * the generated code out of local storage **THIS NEEDS TO
     * BE REVISED - CODES SHOULD BE STORED IN THE DATABASE**
     */
    hidePassword() {
        this.setState({
            visible: false,
        });
        if (localStorage.getItem("originalCode") !== null) {
            localStorage.removeItem("originalCode");
        }
    }

    /**
     * Displays an edit window to the user so they can
     * either update their email address or choose to
     * delete their account
     *
     * @param {Object} e - event object
     */
    onClick(e) {
        const { user } = this.props;
        switch (e.target.id) {
            case "edit-email-btn":
                this.setState({
                    displayEmailWindow: true,
                });
                setTimeout(() => {
                    this.emailField.current.value = user.email;
                }, 500);
                break;
            case "delete-account-btn":
                this.setState({
                    displayDeleteWindow: true,
                    deleteCode: Math.floor(Math.random() * 1000000) + 423523,
                });
                break;
            default:
                return;
        }
    }

    /**
     * Sets the <[e.target.name]> state property to the
     * value of the event target
     *
     * @param {Object} e - event object
     */
    onChange(e) {
        this.setState({
            [e.target.name]: e.target.value,
        });
    }

    /**
     * Updates the user's email address in the database
     *
     * @param {Object} e - event object
     */
    onSubmit(e) {
        e.preventDefault();
        this.setState({
            displayEmailWindow: false,
        });
        const { user, updateUserEmail } = this.props;
        const newAccountState = {
            ...user,
            email: this.state.email,
        };
        updateUserEmail(newAccountState);
    }

    /**
     * Sets the 'visible' state property to true to display
     * the verification code entry field to the user. Also
     * sends an email to the user with the code they have
     * to enter to change their password
     */
    async sendChangeCode() {
        this.setState({
            visible: true,
        });
        const { user } = this.props;
        const res = await sendPWEmail(user.email);
        setTimeout(() => {
            localStorage.setItem("originalCode", res.data.originalCode);
        }, 500);
    }

    /**
     * Checks the verification code the user entered
     * against the original code being stored in
     * local storage. If the codes match, the user is
     * give an auth token and redirected to a protected
     * page where they can update their password. If the
     * codes don't match, the user is shown an error
     * message
     *
     * @param {Object} e - event object
     */
    async submitCode(e) {
        e.preventDefault();
        this.displaySpinner();
        const { confirmCode } = this.state;
        const data = {
            originalCode: localStorage.getItem("originalCode"),
            confirmCode,
            forgottenDetails: false,
        };
        const res = await verifyPWCode(data);
        if (res.success) {
            await this.props.loadUser();
            localStorage.removeItem("originalCode");
            this.setState({
                failed: false,
                authenticated: true,
            });
            this.props.history.push("/update-credentials");
        } else {
            this.setState({
                failed: true,
                verifyFailedMsg: res.msg,
            });
            this.removeSpinner();
        }
    }

    /**
     * Checks to see if the code displayed to the user
     * when the window opens matches the code the user
     * enters into the input field. If it matches, the
     * user's account is deleted and they are redirected
     * to the homepage. If not, an error message is shown
     *
     * @param {Object} e - event object
     */
    onDelete(e) {
        e.preventDefault();
        const { user, history, deleteUser } = this.props;
        const { deleteCode, confirmDeleteCode } = this.state;
        if (deleteCode === Number(confirmDeleteCode)) {
            history["from"] = "delete";
            history.push("/redirect");
            deleteUser(user.email);
        } else {
            this.setState({
                deleteCodeIncorrect: true,
            });
            setTimeout(() => {
                this.setState({
                    deleteCodeIncorrect: false,
                });
            }, 3000);
        }
    }

    render() {
        const {
            displayEmailWindow,
            displayDeleteWindow,
            deleteCode,
            deleteCodeIncorrect,
            visible,
            authenticated,
            failed,
            loading,
        } = this.state;

        return (
            <main className='profile-container'>
                <MobileNav />
                <SidePanel />
                <div className='user-info-container'>
                    <section className='account'>
                        <h2 className='profile-heading' id='account-h2'>
                            My Account
                        </h2>
                        <AccountInfo
                            user={this.props.user}
                            displayEmailWindow={displayEmailWindow}
                            onClick={this.onClick}
                            cancelEdit={this.cancelEdit}
                            onChange={this.onChange}
                            onSubmit={this.onSubmit}
                            emailField={this.emailField}
                        />
                    </section>
                    <section className='password'>
                        <h2 className='profile-heading' id='password-h2'>
                            Password Settings
                        </h2>
                        <Password
                            visible={visible}
                            authenticated={authenticated}
                            displayPasswordChange={this.displayPasswordChange}
                            hidePassword={this.hidePassword}
                            onChange={this.onChange}
                            sendChangeCode={this.sendChangeCode}
                            submitCode={this.submitCode}
                            failed={failed}
                            verifyFailed={this.verifyFailed}
                            loading={loading}
                        />
                    </section>
                    <section className='privacy'>
                        <h2 className='profile-heading' id='privacy-h2'>
                            Privacy
                        </h2>
                        <section className='profile-form-container'>
                            <div className='profile-box'>
                                <p className='profile-label'>
                                    Profile visibility:
                                </p>
                                <div className='profile-visibility'>
                                    <input
                                        type='checkbox'
                                        id='profile-public'
                                        name='profile-public'
                                        defaultChecked
                                    />
                                    <label htmlFor='profile-public'>
                                        Public
                                    </label>
                                    <br />
                                    <input
                                        type='checkbox'
                                        id='profile-private'
                                        name='profile-private'
                                    />
                                    <label htmlFor='profile-private'>
                                        Private
                                    </label>
                                </div>
                            </div>
                            <div className='profile-box'>
                                <p className='profile-label'>
                                    Receive emails from us?
                                </p>
                                <div className='profile-visibility'>
                                    <input
                                        type='checkbox'
                                        id='receive-email-pos'
                                        name='receive-email-pos'
                                        defaultChecked
                                    />
                                    <label htmlFor='receive-email-pos'>
                                        Yes
                                    </label>
                                    <br />
                                    <input
                                        type='checkbox'
                                        id='receive-email-neg'
                                        name='receive-email-neg'
                                    />
                                    <label htmlFor='receive-email-neg'>
                                        No
                                    </label>
                                </div>
                            </div>
                        </section>
                    </section>
                    <section className='delete-account-container'>
                        <DeleteAccount
                            onClick={this.onClick}
                            onChange={this.onChange}
                            cancelEdit={this.cancelEdit}
                            displayDeleteWindow={displayDeleteWindow}
                            deleteField={this.deleteField}
                            onDelete={this.onDelete}
                            deleteCode={deleteCode}
                            deleteCodeIncorrect={deleteCodeIncorrect}
                        />
                    </section>
                </div>
            </main>
        );
    }
}

Account.propTypes = {
    user: PropTypes.object,
    displayAlert: PropTypes.func,
    updateUserEmail: PropTypes.func.isRequired,
    loadUser: PropTypes.func,
    deleteUser: PropTypes.func,
};

const setProps = (state) => ({
    user: state.auth.user,
});

export default connect(setProps, {
    displayAlert,
    updateUserEmail,
    deleteUser,
    loadUser,
})(withRouter(Account));
