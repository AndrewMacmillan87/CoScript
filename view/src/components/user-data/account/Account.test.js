import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../../redux/store/reduxStore";

import { Account } from "./Account";
import MobileNav from "../MobileNav";
import SidePanel from "../SidePanel";
import AccountInfo from "./AccountInfo";
import Password from "./Password";
import DeleteAccount from "./DeleteAccount";
import EditAccountData from "./EditAccountData";
import Button from "../../buttons/Button";

describe("Account component", () => {
    let props, mockUpdateUserEmail, mockDeleteUser;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockUpdateUserEmail = jest.fn();
        mockDeleteUser = jest.fn();
        props = {
            user: { email: "test@email.com" },
            updateUserEmail: mockUpdateUserEmail,
            deleteUser: mockDeleteUser,
            history: {
                from: "",
                push: jest.fn(),
            },
        };
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Account {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a main element with a className of 'profile-container'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(".profile-container").length).toEqual(1);
    });

    it("should render a div element with a className of 'user-info-container'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(".user-info-container").length).toEqual(1);
    });

    it("should render 7 section elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(7);
    });

    it("should render a section element with a className of 'account'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(".account").length).toEqual(1);
    });

    it("should render a section element with a className of 'password'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(".password").length).toEqual(1);
    });

    it("should render a section element with a className of 'privacy'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(".privacy").length).toEqual(1);
    });

    it("should render 3 section elements with a className of 'profile-form-container'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(".profile-form-container").length).toEqual(3);
    });

    it("should render 3 h2 elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("h2").length).toEqual(3);
    });

    it("should render 5 label elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("label").length).toEqual(5);
    });

    it("should render 4 input elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("input").length).toEqual(4);
    });

    it("should render the MobileNav component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(MobileNav).length).toEqual(1);
    });

    it("should render the SidePanel component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(SidePanel).length).toEqual(1);
    });

    it("should render the AccountInfo component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(AccountInfo).length).toEqual(1);
    });

    it("should render the Password component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Password).length).toEqual(1);
    });

    it("should render the DeleteAccount component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(DeleteAccount).length).toEqual(1);
    });

    it("should call `updateUserEmail()` when the email change form is submitted", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        wrapper.find(Account).setState({ displayEmailWindow: true });
        wrapper
            .find("input")
            .at(5)
            .simulate("change", {
                target: { name: "email", value: "newemail@test.com" },
            });
        wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
        expect(mockUpdateUserEmail).toBeCalled();
    });

    it("should call `deleteUser()` when the delete form is submitted", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...props} />
                </Router>
            </Provider>
        );
        wrapper.find(Account).setState({
            displayDeleteWindow: true,
            deleteCode: 123456,
            confirmDeleteCode: "123456",
        });
        const deleteWrapper = wrapper.find(DeleteAccount);
        deleteWrapper
            .find("form")
            .at(0)
            .simulate("submit", { preventDefault: jest.fn() });
        await expect(mockDeleteUser).toBeCalled();
    });
});

describe("AccountInfo component", () => {
    let accountProps, props, mockUpdateUserEmail, mockDeleteUser;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockUpdateUserEmail = jest.fn();
        mockDeleteUser = jest.fn();
        accountProps = {
            user: { email: "test@email.com" },
            updateUserEmail: mockUpdateUserEmail,
            deleteUser: mockDeleteUser,
            history: {
                from: "",
                push: jest.fn(),
            },
        };
        props = {
            displayEmailWindow: false,
            onSubmit: jest.fn(),
            onClick: jest.fn(),
            cancelEdit: jest.fn(),
            onChange: jest.fn(),
            emailField: {},
        };
    });

    it("should render the EditAccountData component when 'displayEmailWindow' is true", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...accountProps} />
                </Router>
            </Provider>
        );
        wrapper.find(Account).setState({ displayEmailWindow: true });
        const accountInfoWrapper = wrapper.find(AccountInfo);
        expect(accountInfoWrapper.find(EditAccountData).length).toEqual(1);
    });

    it("should render a section element with a className of 'profile-form-container'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <AccountInfo {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(1);
        expect(wrapper.find(".profile-form-container").length).toEqual(1);
    });

    it("should render 3 div elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <AccountInfo {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("div").length).toEqual(3);
    });

    it("should render a button element", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <AccountInfo {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("button").length).toEqual(1);
    });

    it("should render an image element", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <AccountInfo {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("img").length).toEqual(1);
    });
});

describe("DeleteAccount component", () => {
    let accountProps, props, mockUpdateUserEmail, mockDeleteUser;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockUpdateUserEmail = jest.fn();
        mockDeleteUser = jest.fn();
        accountProps = {
            user: { email: "test@email.com" },
            updateUserEmail: mockUpdateUserEmail,
            deleteUser: mockDeleteUser,
            history: {
                from: "",
                push: jest.fn(),
            },
        };
        props = {
            user: {},
            onClick: jest.fn(),
            onChange: jest.fn(),
            cancelEdit: jest.fn(),
            onDelete: jest.fn(),
            displayDeleteWindow: false,
            deleteField: {},
            deleteCode: 0,
            deleteCodeIncorrect: false,
        };
    });

    it("should render the EditAccountData component when 'displayEmailWindow' is true", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Account {...accountProps} />
                </Router>
            </Provider>
        );
        wrapper.find(Account).setState({ displayEmailWindow: true });
        const deleteAccountWrapper = wrapper.find(AccountInfo);
        expect(deleteAccountWrapper.find(EditAccountData).length).toEqual(1);
    });

    it("should render a h3 element with a className of 'delete-heading'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <DeleteAccount {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("h3").length).toEqual(1);
        expect(wrapper.find(".delete-heading").length).toEqual(1);
    });

    it("should render 2 p elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <DeleteAccount {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").length).toEqual(2);
    });

    it("should render the Button component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <DeleteAccount {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Button).length).toEqual(1);
    });
});

describe("Password component", () => {
    let props, accountProps;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        accountProps = {
            user: { email: "test@email.com" },
            updateUserEmail: jest.fn(),
            history: {
                from: "",
                push: jest.fn(),
            },
        };
        props = {
            visible: false,
            onChange: jest.fn(),
            sendChangeCode: jest.fn(),
            submitCode: jest.fn(),
            hidePassword: jest.fn(),
            verifyFailed: {},
            failed: false,
            loading: false,
        };
    });

    describe("Form not visible", () => {
        it("should render a section element with a className of 'profile-form-container'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Password {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("section").length).toEqual(1);
            expect(wrapper.find(".profile-form-container").length).toEqual(1);
        });

        it("should render a div element with a className of 'pw-button-container'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Password {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("div").length).toEqual(1);
            expect(wrapper.find(".pw-button-container").length).toEqual(1);
        });

        it("should render 1 Button component", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Password {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find(Button).length).toEqual(1);
        });
    });

    describe("Form is visible", () => {
        it("should render 2 div elements with a className of 'profile-box'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Account {...accountProps} />
                    </Router>
                </Provider>
            );
            wrapper.find(Account).setState({ visible: true });
            const password = wrapper.find(Password);
            expect(password.find(".profile-box").length).toEqual(2);
        });

        it("should render 3 button elements", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Account {...accountProps} />
                    </Router>
                </Provider>
            );
            wrapper.find(Account).setState({ visible: true });
            const password = wrapper.find(Password);
            expect(password.find("button").length).toEqual(3);
        });

        it("should render a div element with a className of '.pw-code-container'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Account {...accountProps} />
                    </Router>
                </Provider>
            );
            wrapper.find(Account).setState({ visible: true });
            const password = wrapper.find(Password);
            expect(password.find(".pw-code-container").length).toEqual(1);
        });

        it("should render an input element with a className of '.pw-code-input'", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Account {...accountProps} />
                    </Router>
                </Provider>
            );
            wrapper.find(Account).setState({ visible: true });
            const password = wrapper.find(Password);
            expect(password.find(".pw-code-input").length).toEqual(1);
        });

        it("should render an img with a className of '.verify-pw-code-spinner' when 'loading' is true", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Account {...accountProps} />
                    </Router>
                </Provider>
            );
            wrapper.find(Account).setState({ visible: true, loading: true });
            const password = wrapper.find(Password);
            expect(password.find(".verify-pw-code-spinner").length).toEqual(1);
        });

        it("should render an img with a className of '.pw-code-failed' when 'failed' is true", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <Account {...accountProps} />
                    </Router>
                </Provider>
            );
            wrapper.find(Account).setState({ visible: true, failed: true });
            const password = wrapper.find(Password);
            expect(password.find(".pw-code-failed").length).toEqual(1);
        });
    });
});
