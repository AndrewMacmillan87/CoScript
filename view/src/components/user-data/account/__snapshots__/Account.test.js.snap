// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account component renders correctly 1`] = `
<main
  className="profile-container"
>
  <nav
    className="profile-mobile-nav"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <i
        className="home-icon fas fa-home fa-3x"
      />
    </a>
    <button
      className="profile-dropdown-icon-link"
      onClick={[Function]}
    >
      <img
        alt="Burger menu icon"
        className="profile-dropdown-icon"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "menu.svg",
          }
        }
      />
    </button>
    <div
      className="dropdown-menu dropdown-content"
      id="dropdown-nav"
    >
      <a
        href="/profile"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          My Profile
        </p>
      </a>
      <a
        href="/messages"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Messages
        </p>
      </a>
      <a
        href="/projects"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Active Projects
        </p>
      </a>
      <a
        href="/publications"
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Publications
        </p>
      </a>
      <a
        href=" "
        onClick={[Function]}
      >
        <p
          className="dropdown-link"
        >
          Logout
        </p>
      </a>
    </div>
  </nav>
  <div
    className="user-info-container"
  >
    <section
      className="account"
    >
      <h2
        className="profile-heading"
        id="account-h2"
      >
        My Account
      </h2>
      <section
        className="profile-form-container"
      >
        <div
          className="profile-box"
        >
          <label
            className="profile-label"
            htmlFor="user-email"
          >
            Email Address:
          </label>
          <div
            className="input-container"
          >
            <div
              className="account-box"
            >
              <p
                className="users-email"
              >
                test@email.com
              </p>
            </div>
            <button
              className="edit-icon-btn"
              onClick={[Function]}
            >
              <img
                alt="edit input icon"
                className="edit-input-icon"
                id="edit-email-btn"
                src={
                  Object {
                    "ReactComponent": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "render": [Function],
                    },
                    "__esModule": true,
                    "default": "edit.svg",
                  }
                }
              />
            </button>
          </div>
        </div>
      </section>
    </section>
    <section
      className="password"
    >
      <h2
        className="profile-heading"
        id="password-h2"
      >
        Password Settings
      </h2>
      <section
        className="profile-form-container"
      >
        <div
          className="pw-button-container"
        >
          <button
            className="small-btn"
            onClick={[Function]}
            type="button"
          >
            <span
              className="btn-text"
            >
              <b>
                Change Password
              </b>
            </span>
          </button>
        </div>
      </section>
    </section>
    <section
      className="privacy"
    >
      <h2
        className="profile-heading"
        id="privacy-h2"
      >
        Privacy
      </h2>
      <section
        className="profile-form-container"
      >
        <div
          className="profile-box"
        >
          <p
            className="profile-label"
          >
            Profile visibility:
          </p>
          <div
            className="profile-visibility"
          >
            <input
              defaultChecked={true}
              id="profile-public"
              name="profile-public"
              type="checkbox"
            />
            <label
              htmlFor="profile-public"
            >
              Public
            </label>
            <br />
            <input
              id="profile-private"
              name="profile-private"
              type="checkbox"
            />
            <label
              htmlFor="profile-private"
            >
              Private
            </label>
          </div>
        </div>
        <div
          className="profile-box"
        >
          <p
            className="profile-label"
          >
            Receive emails from us?
          </p>
          <div
            className="profile-visibility"
          >
            <input
              defaultChecked={true}
              id="receive-email-pos"
              name="receive-email-pos"
              type="checkbox"
            />
            <label
              htmlFor="receive-email-pos"
            >
              Yes
            </label>
            <br />
            <input
              id="receive-email-neg"
              name="receive-email-neg"
              type="checkbox"
            />
            <label
              htmlFor="receive-email-neg"
            >
              No
            </label>
          </div>
        </div>
      </section>
    </section>
    <section
      className="delete-account-container"
    >
      <h3
        className="delete-heading"
      >
        Delete your account?
      </h3>
      <p>
        <b>
          Warning
        </b>
      </p>
      <p
        className="delete-disclaimer"
      >
        You cannot undo this action and all documents associated with your account will be permanently deleted. Please only delete your account after ensuring any work you have saved on the system is safely backed up. Be aware that all master copies of projects you created will be sent to all of the collaborators on those projects when you delete your account
      </p>
      <button
        className="small-btn"
        id="delete-account-btn"
        onClick={[Function]}
        type="button"
      >
        <span
          className="btn-text"
        >
          <b>
            Delete
          </b>
        </span>
      </button>
    </section>
  </div>
</main>
`;
